### PR TITLE
wallet: #25768 follow ups

### PR DESCRIPTION
**Pull request by stickies-v**
This PR addresses the outstanding comments/issues from #25768:

- capitalization [typo](https://github.com/bitcoin/bitcoin/pull/25768#discussion_r958572522) in docstring
- remove [unused locks](https://github.com/bitcoin/bitcoin/commit/01f3534632d18c772901fb6ce22f6394eae96799) that we previously needed for `ReacceptWalletTransactions()`
- before #25768, only `ResendWalletTransactions()` would reset `m_next_resend` (formerly called `nNextResend`). By unifying it with `ReacceptWalletTransactions()` into `ResubmitWalletTransactions()`, the number of callsites that would reset the `m_next_resend` timer increased
  - since `m_next_resend` is only used in case of `relay=true` (formerly `ResendWalletTransactions()`), this is unintuitive
  - it leads to [unexpected behaviour](https://github.com/bitcoin/bitcoin/pull/25768#issuecomment-1252619427) such as transactions potentially never being rebroadcasted.
  - it makes the ResubmitWalletTransactions()` logic [more complicated than strictly necessary](https://github.com/bitcoin/bitcoin/pull/25768#discussion_r962828563)
  - since #25768, we relied on an earlier call of `ResubmitWalletTransactions(relay=false, force=true)` to initialize `m_next_resend()`, I think we can more elegantly do that by just providing `m_next_resend` with a default value
  - just to highlight: this commit introduces behaviour change

Note: the `if (!fBroadcastTransactions)` in `CWallet:ShouldResend()` is duplicated on purpose, since it potentially avoids the slightly more expensive `if (!chain().isReadyToBroadcast())` check afterwards. I don't have a strong view on it, so happy to remove that additional check to reduce the diff, too.

### CODE REVIEW

File: src/wallet/wallet.h [hunk 1]
```diff
@@ -537,6 +540,10 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
     };
     ScanResult ScanForWalletTransactions(const uint256& start_block, int start_height, std::optional<int> max_height, const WalletRescanReserver& reserver, bool fUpdate, const bool save_progress);
     void transactionRemovedFromMempool(const CTransactionRef& tx, MemPoolRemovalReason reason, uint64_t mempool_sequence) override;
+    /** Set the next time this wallet should resend transactions, or use the default. */
+    void SetNextResend(int next_resend = GetDefaultNextResend()) { m_next_resend = next_resend; }
```
Comment (MarcoFalke) [inline src/wallet/wallet.h:544]: Any reason to add a function that is only called once?

File: src/wallet/wallet.h [hunk 2]
```diff
@@ -348,6 +349,8 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
      */
     static bool AttachChain(const std::shared_ptr<CWallet>& wallet, interfaces::Chain& chain, const bool rescan_required, bilingual_str& error, std::vector<bilingual_str>& warnings);
 
+    static int GetDefaultNextResend() { return GetTime() + (12 * 60 * 60) + GetRand(24 * 60 * 60); }
```
Comment (MarcoFalke) [inline src/wallet/wallet.h:352]: Any reason to truncate the result?

File: src/wallet/wallet.h [hunk 3]
```diff
@@ -250,7 +251,7 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
     int nWalletVersion GUARDED_BY(cs_wallet){FEATURE_BASE};
 
     /** The next scheduled rebroadcast of wallet transactions. */
-    std::atomic<int64_t> m_next_resend{};
+    std::atomic<int64_t> m_next_resend{GetDefaultNextResend()};
```
Comment (MarcoFalke) [inline src/wallet/wallet.h:254]: No longer needs to be atomic?

File: src/wallet/wallet.cpp [hunk 4]
```diff
@@ -1979,7 +1984,9 @@ void CWallet::ResubmitWalletTransactions(bool relay, bool force)
 void MaybeResendWalletTxs(WalletContext& context)
 {
     for (const std::shared_ptr<CWallet>& pwallet : GetWallets(context)) {
+        if (!pwallet->ShouldResend()) continue;
         pwallet->ResubmitWalletTransactions(/*relay=*/true, /*force=*/false);
+        pwallet->SetNextResend();
```
Comment (MarcoFalke) [inline src/wallet/wallet.cpp:1989]: Might be good to have bugfixes in separate commits from refactoring changes

**Review by maflcko — APPROVED** [review]

File: src/wallet/wallet.h (hunk 1 above)
Comment (stickies-v) [reply src/wallet/wallet.h:544]: Because `m_next_resend` is a private member, so I figured we'd prefer using a setter instead of making the variable public? Happy to take the other approach though, no strong preference.

File: src/wallet/wallet.h (hunk 2 above)
Comment (stickies-v) [reply src/wallet/wallet.h:352]: What do you mean with truncating here? I just moved the existing logic into a new function, that on its own shouldn't be any behavior change.

File: src/wallet/wallet.h (hunk 1 above)
Comment (MarcoFalke) [reply src/wallet/wallet.h:544]: Sounds good. If you keep the function, the argument can be removed. If you keep the argument, the type needs to be adjusted to be not truncating.

**Review by stickies-v — COMMENTED** [review]
Force pushed to address MarcoFalke's review comments - thank you!

- Split out "wallet: move resend timer logic out of ResubmitWalletTransactions" into two commits: one refactoring commit (that introduces `ShouldResend()` but has no behaviour change) and one bugfix commit that only updates `m_next_resend` when we're actually resending the transactions
- Fixed int truncation bug
- No need for `m_next_resend` to be `std::atomic` anymore, since it's now only called from a single thread (`MaybeResendWalletTxs()`)
- Made `SetNextResend()` argument-less since we're never actually passing the optional argument, can easily add back later.

File: src/wallet/wallet.h [hunk 5]
```diff
@@ -348,6 +349,8 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
      */
     static bool AttachChain(const std::shared_ptr<CWallet>& wallet, interfaces::Chain& chain, const bool rescan_required, bilingual_str& error, std::vector<bilingual_str>& warnings);
 
+    static int64_t GetDefaultNextResend() { return GetTime() + (12 * 60 * 60) + GetRand(24 * 60 * 60); }
```
Comment (MarcoFalke) [inline src/wallet/wallet.h:352]: nit: Could put the impl into the cpp file to avoid a header include in the header?

**Review by maflcko — APPROVED** [review]
lgtm. I wouldn't call the last commit a "performance" improvement. Just a code clarity change.

File: src/wallet/wallet.cpp [hunk 6]
```diff
@@ -1934,17 +1953,6 @@ void CWallet::ResubmitWalletTransactions(bool relay, bool force)
...
-    // During reindex, importing and IBD, old wallet transactions become
-    // unconfirmed. Don't resend them as that would spam other nodes.
-    // We only allow forcing mempool submission when not relaying to avoid this spam.
-    if (!force && relay && !chain().isReadyToBroadcast()) return;
-
-    // Do this infrequently and randomly to avoid giving away
-    // that these are our transactions.
-    if (!force && GetTime() < m_next_resend) return;
-    // resend 12-36 hours from now, ~1 day on average.
-    m_next_resend = GetTime() + (12 * 60 * 60) + GetRand(24 * 60 * 60);
```
Comment (aureleoules) [inline src/wallet/wallet.cpp:1946]: By removing this line, calls such as `importaddress`, `importpubkey`, `importmulti` and `importdescriptors` with rescan enabled won't update the timeout because now the timeout is only reset in `MaybeResendWalletTxs` with `SetNextResend`, is this normal?

See https://github.com/bitcoin/bitcoin/blob/6610cc99a6a479ca22d272a1d583ab2a8422b7ae/src/wallet/rpc/backup.cpp#L471-L475

**Review by aureleoules — COMMENTED** [review]
01f3534632d18c772901fb6ce22f6394eae96799: I verified that `ResubmitWalletTransactions` already locks the mutex so no need to lock it in the parent functions.

Left a comment, the rest looks good.

File: src/wallet/wallet.cpp (hunk 6 above)
Comment (stickies-v) [reply src/wallet/wallet.cpp:1946]: Yes, this is intentional. Prior to #25768, we had two functions [`ReacceptWalletTransactions()`](https://github.com/achow101/bitcoin/blob/e191fac4f3c37820f0618f72f0a8e8b524531ab8/src/wallet/wallet.cpp#L1860) (which did not update/read `nNextResend`) and [`ResendWalletTransactions()`](https://github.com/achow101/bitcoin/blob/e191fac4f3c37820f0618f72f0a8e8b524531ab8/src/wallet/wallet.cpp#L1937) (which did update/read `nNextResend`, only called from `MaybeUpdateWalletTxs()`), which got merged into a single parameterized `ResubmitWalletTransactions()` (and updates/reads `m_next_resend`).

Since the purpose of `m_next_resend` is to prevent overly frequent _relaying_ of transactions, we should only update the timer when relaying is triggered, so this PR just undoes that behaviour change introduced in #25768.

I tried to summarize this rationale in the commit message of 3e5d3f8705154e09ecf3639f6d261595992fc5b5 - lmk if that message is unclear?

File: src/wallet/wallet.cpp [hunk 7]
```diff
@@ -1903,6 +1903,23 @@ std::set<uint256> CWallet::GetTxConflicts(const CWalletTx& wtx) const
     return result;
 }
 
+bool CWallet::ShouldResend()
```
[the remark this answers is not in the document]
Comment (stickies-v) [reply src/wallet/wallet.cpp:1907]: Good catch, thx. Used to have side effects in `ShouldResend()` in an earlier implementation and forgot to remove.

**Comment by stickies-v** [comment]
Force pushed to make `CWallet::ShouldResend()` `const` - ty aureleoules!

File: src/wallet/wallet.cpp (hunk 6 above)
Comment (aureleoules) [reply src/wallet/wallet.cpp:1946]: Great, that makes sense since these RPC calls don't relay transactions. Your commit message is also clear with this in mind :+1:.

**Comment by aureleoules** [comment]
ACK b01682a812f0841170657708ef0e896b904fcd77

**Comment by luke-jr** [comment]
>I verified that ResubmitWalletTransactions already locks the mutex so no need to lock it in the parent functions.

That's not necessarily a given. The parents may need to lock things for atomicity (with other changes/calls), or lock ordering consistency. (I didn't look at this particular case, just noting the general logic does not hold)

**Comment by luke-jr** [comment]
>it leads to https://github.com/bitcoin/bitcoin/pull/25768#issuecomment-1252619427 such as transactions potentially never being rebroadcasted.

I think this bugfix should be split out and backported. Ideally doing something more minimal which can be refactored here.

File: src/wallet/wallet.cpp (hunk 4 above)
Comment (luke-jr) [reply src/wallet/wallet.cpp:1992]: The "bugfix" commit is full of refactoring still

Comment (luke-jr) [reply src/wallet/wallet.cpp:1992]: Also, cherry-picking the bugfix commit by itself is broken: it will reset `m_next_resend` to 12-36 hours, *every minute*, even if nothing is done/rebroadcast.

**Comment by achow101** [comment]
ACK b01682a812f0841170657708ef0e896b904fcd77

**Comment by fanquake** [comment]
Added to #26133 for backport.
